### PR TITLE
feat: optionally support gasless transaction

### PR DESCRIPTION
**Pull request by RetricSu**
- [x] fetch entrypoint contract config from godwoken rpc
- [x] docs about how to use gasless transactions in additional features.

### CODE REVIEW

File: packages/api-server/src/methods/validator.ts [hunk 1]
```diff
@@ -662,6 +664,147 @@ export async function verifyEnoughBalance(
...
+
+  // extra check
+  // 1. no base fee since we are not impl eip1559
+  if (userOperation.maxFeePerGas != userOperation.maxFeePerGas) {
+    return invalidParamsError(
+      index,
+      `userOperation.maxFeePerGas(${userOperation.maxFeePerGas}) != userOperation.maxFeePerGas(${userOperation.maxFeePerGas})`
+    );
+  }
+  // 2. gasLimit = verificationGasLimit * 3 + callGasLimit
```
Comment (sopium) [inline packages/api-server/src/methods/validator.ts:782]: Ignorant question: why should `gasLimit = verificationGasLimit * 3 + callGasLimit`? Specifically, why `* 3`?

Comment (RetricSu) [reply packages/api-server/src/methods/validator.ts:782]: https://github.com/godwokenrises/account-abstraction/blob/gw-gasless/contracts/core/GaslessEntryPoint.sol#L327-L338

similar to https://github.com/godwokenrises/account-abstraction/blob/develop/contracts/core/EntryPoint.sol#L265-L268

File: packages/api-server/src/gasless/payload.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
+import { HexNumber, HexString } from "@ckb-lumos/base";
+import { AbiCoder } from "web3-eth-abi";
+const abiCoder = require("web3-eth-abi") as AbiCoder;
```
Comment (jjyr) [inline packages/api-server/src/gasless/payload.ts:3]: Override the previous import?

File: packages/api-server/src/methods/modules/gw.ts [hunk 3]
```diff
@@ -502,6 +500,37 @@ export class Gw {
           throw intrinsicGasErr.padContext(`gw_submit_l2transaction`);
         }
 
+        // check gas price
+        const isZeroGasPrice = gasPrice == "0x" || gasPrice == "0x0";
+        if (isZeroGasPrice) {
+          if (entrypointContract == null) {
+            throw new Error(
```
Comment (jjyr) [inline packages/api-server/src/methods/modules/gw.ts:507]: We should accept zero gas tx unless if can't pass minimal gas price check

File: packages/api-server/src/gasless/payload.ts (hunk 2 above)
Comment (RetricSu) [reply packages/api-server/src/gasless/payload.ts:3]: the library `web3-eth-abi` does not works very well with typescript, here is a simple workaround

File: packages/api-server/src/convert-tx.ts [hunk 4]
```diff
@@ -367,11 +369,33 @@ export async function polyTxToGwTx(
...
+      gasPrice === "0x" ? "0x0" : gasPrice,
+      gasLimit === "0x" ? "0x0" : gasLimit,
+      0
+    );
+    if (err != null) {
+      throw err.padContext(`eth_sendRawTransaction ${polyTxToGwTx.name}`);
+    }
+  }
+
+  // Check gas price
```
Comment (magicalne) [inline packages/api-server/src/convert-tx.ts:398]: We should not check gas price again for `0 gas price` if it's a gasless tx.

**Review by magicalne — CHANGES_REQUESTED** [review]

File: packages/api-server/src/convert-tx.ts (hunk 4 above)
Comment (RetricSu) [reply packages/api-server/src/convert-tx.ts]: updated

**Review by magicalne — APPROVED** [review]

**Review by jjyr — APPROVED** [review]

File: packages/api-server/src/methods/modules/poly.ts [hunk 5]
```diff
@@ -89,6 +89,10 @@ export class Poly {
         backends: gwConfig.backends,
         accounts: gwConfig.accounts,
         chainId: gwConfig.web3ChainId,
+        gaslessTx: {
+          support: gwConfig.entrypointContract ? true : false,
```
Comment (classicalliu) [inline packages/api-server/src/methods/modules/poly.ts:93]:
```suggestion
          support: !!gwConfig.entrypointContract,
```

File: packages/api-server/src/methods/validator.ts [hunk 6]
```diff
@@ -662,6 +664,153 @@ export async function verifyEnoughBalance(
...
+  gasLimit: HexNumber,
+  index: number
+) {
+  if (gwConfig.entrypointContract == null) {
+    throw new Error(
+      "can not call verifyGaslessTransaction when entrypoint contract is null"
+    );
+  }
+
+  if (gasPrice != "0x0") {
```
Comment (classicalliu) [inline packages/api-server/src/methods/validator.ts:681]:
```suggestion
  if (gasPrice !== "0x0") {
```

File: packages/api-server/src/methods/validator.ts [hunk 7]
```diff
@@ -662,6 +664,153 @@ export async function verifyEnoughBalance(
...
+  // 1. no base fee since we are not impl eip1559
+  if (userOperation.maxFeePerGas != userOperation.maxFeePerGas) {
+    return invalidParamsError(
+      index,
+      `userOperation.maxFeePerGas(${userOperation.maxFeePerGas}) != userOperation.maxFeePerGas(${userOperation.maxFeePerGas})`
+    );
+  }
+  // 2. gasLimit = verificationGasLimit * 3 + callGasLimit
+  if (
+    BigInt(gasLimit) !=
```
Comment (classicalliu) [inline packages/api-server/src/methods/validator.ts:790]:
```suggestion
    BigInt(gasLimit) !==
```

File: packages/api-server/src/methods/validator.ts [hunk 8]
```diff
@@ -662,6 +664,153 @@ export async function verifyEnoughBalance(
...
+        10
+      )}) != verificationGasLimit(${BigInt(
+        userOperation.verificationGasLimit
+      ).toString(10)}) * 3 + callGasLimit(${BigInt(
+        userOperation.callGasLimit
+      ).toString(10)})`
+    );
+  }
+  // 3. to address should equal entrypoint
+  if (to != gwConfig.entrypointContract.address) {
```
Comment (classicalliu) [inline packages/api-server/src/methods/validator.ts:806]:
```suggestion
  if (to !== gwConfig.entrypointContract.address) {
```

**Review by classicalliu — APPROVED** [review]

**Pull request merged**